### PR TITLE
Sort help contexts by specificity

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Instead of just sorting the help context alphabetically we now first sort them by context match specificity. Context with equal specificity are sorted alphabetically.

### CODE REVIEW

File: core/help/help.py [hunk 1]
```diff
@@ -438,6 +448,54 @@ def refresh_context_command_map(enabled_only=False):
...
+            return score_b - score_a
+        if short_name_a < short_name_b:
+            return -1
+        if short_name_a > short_name_b:
+            return 1
+        return 0
+
+    return sorted(
+        display_name_to_context_name_map.keys(),
+        key=cmp_to_key(sort_key),
```
Comment (auscompgeek) [inline core/help/help.py:495]:
```suggestion
        key=get_specificity_score,
```

Comment (AndreasArvidsson) [reply core/help/help.py:495]: Now you're not sorting alphabetically anymore.

Comment (auscompgeek) [reply core/help/help.py:495]: Fair.
```suggestion
        key=lambda name: (get_specificity_score(name), name),
```

Comment (AndreasArvidsson) [reply core/help/help.py:495]: I didn't know you could use a topple like that. Thanks. 
The problem is still that we want specificity to be sorted in descending order. Name should still be in ascending order.

Comment (auscompgeek) [reply core/help/help.py:495]: and this is why I hate `cmp` lol
```suggestion
        key=lambda name: (-get_specificity_score(name), name),
```

Comment (AndreasArvidsson) [reply core/help/help.py:495]: This actually recreates the same behavior with less code. Excellent!

**Comment by nriley** [comment]
We discussed this on the community backlog session today. @AndreasArvidsson and I felt that the current version is essentially a proof of concept and would not be mergeable in its current state. Some potential next steps:

- Figure out some heuristics that create fewer alphabetized sections, at least for stock community matchers, and potentially add separators (lines/blank space) between the alphabetized sections
- Stop using imgui for community's help and instead build something canvas based which would give us a lot more formatting flexibility to be able to explain the richness of the active contexts. See also how we are stuck on #1431 (though we also have another potential short term fix there in splitting it between two different help displays).

If anybody would be interested in working on the first option, feel free to try out/suggest something!

**Comment by AndreasArvidsson** [comment]
What if we group them in three different groups? Application context/with context/global?

**Comment by AndreasArvidsson** [comment]
Something like this

Screenshot 2024-06-23T11-20-06

**Comment by nriley** [comment]
Looks good to me, maybe with some slight wordsmithing "Application-specific:"; "Context-dependent"? We would need to make sure that we don't run over the total # of lines so this doesn't cause help to go off the bottom of the screen.

**Comment by AndreasArvidsson** [comment]
Changing the words is no problem. Updating how max contexts per page is implemented is about bit worse.
Basically this action now needs to be aware of the number of groupings. Not he number of potential groupings, but actual found. 

https://github.com/talonhub/community/blob/3b2e6048fb7832f39cc23ccb0faa83d407a87abc/core/help/help.py#L121-L140

**Comment by AndreasArvidsson** [comment]
Also the setting is number of context per page. Not number of lines per page. I'm very tempted to leave it as it is and the user have to reduce the setting somewhat if the page is to large now. They can always change the setting and go back to the previous behavior if they prefer that?

**Comment by AndreasArvidsson** [comment]
@nriley That pr is now updated with this grouping and your suggested names.

**Comment by nriley** [comment]
Looks great. The only thing that seemed to be an obvious misclassification was text_navigation.talon due to its use of `not app:`:

https://github.com/nriley/talon_community/blob/e11633b3343e25a005dd35991934f6b820f59963/plugin/text_navigation/text_navigation.talon#L1

**Comment by AndreasArvidsson** [comment]
That is harder and requires us to actually parse the context

**Comment by nriley** [comment]
Maybe we can change the matcher for text_navigation.talon then?

**Comment by AndreasArvidsson** [comment]
> Maybe we can change the matcher for text_navigation.talon then?

Probably more filling to use a tag. Can't be only vscode where this should not be active?

**Comment by nriley** [comment]
Yeah, I'm thinking it mostly means "this app has a better way of navigating through text than this", which at the moment means cursorless. Is there something that is getting used consistently by the various non-VSCode cursorless implementations?

**Comment by AndreasArvidsson** [comment]
We do have a Cursorless tag

**Comment by nriley** [comment]
So does that seem reasonable to use — not tag: user.cursorless?

**Comment by AndreasArvidsson** [comment]
Sounds reasonable, but is not application vscode something you added yourself?
https://github.com/talonhub/community/blob/main/plugin/text_navigation/text_navigation.talon

**Comment by nriley** [comment]
Wow, that's embarrassing. Let's go ahead and merge this and I'll clean up my act :-)

**Review by nriley — APPROVED** [review]

**Comment by jaresty** [comment]
Is there any trick to seeing this sort? I tried running 'help context' but I didn't see any changes, but I'm not sure I was looking in the right place 🤔

**Comment by nriley** [comment]
This does sort the output of "help context", but is more useful with "help active".

**Comment by jaresty** [comment]
That's what I was missing. Thank you

**Pull request merged**